### PR TITLE
Add and remove mobile client from service instance

### PR DESCRIPTION
**Pull request by sedroche**
A mobile developer will want to see which clients are associated with which service instances.
This PR adds a component to mobile aware service instances to add and exclude clients from the service.

### Single client
single-client

### Multiple clients
multiple-clients

### Add client button becomes enabled when excluded clients are available
enabled-button

### Add Client modal
add-client-modal

### CODE REVIEW

File: app/scripts/directives/addMobileClient.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        .then(function() {
+          NotificationsService.addNotification({
+            type: "success",
+            message: "Successfully added " + mobileClient.metadata.name + " client."
+          });
+        })
+        .catch(function(error) {
+          NotificationsService.addNotification({
+            type: "error",
+            message: "Failed to add mobile client",
```
Comment (spadgett) [inline app/scripts/directives/addMobileClient.js]: The messages here and below should have punctuation.

"Failed to add mobile client."

File: app/scripts/filters/mobile.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+/* jshint unused: false */
```
Comment (spadgett) [inline app/scripts/filters/mobile.js]: Why is this needed?

File: app/scripts/directives/mobileServiceClients.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+    ctrl.showOverlayPanel = function(panelName, state) {
+      _.set(ctrl, 'overlay.panelVisible', true);
+      _.set(ctrl, 'overlay.panelName', panelName);
+      _.set(ctrl, 'overlay.state', state);
+    };
+
+    ctrl.canAddMobileClient = function() {
+      return !MobileClientsService.filterExcluded(ctrl.serviceInstance, ctrl.mobileClients).length;
```
Comment (spadgett) [inline app/scripts/directives/mobileServiceClients.js]: nit: prefer !_.isEmpty(...)

File: app/scripts/services/mobileClients.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  .factory("MobileClientsService", function(DataService) {
+
+    var mobileclientVersion = {
+      group: "mobile.k8s.io",
+      version: "v1alpha1",
+      resource: "mobileclients"
+    };
+
+
+    var watch = function(context, callback, opts ) {
```
Comment (spadgett) [inline app/scripts/services/mobileClients.js]: nit: extra space before closing paren

File: app/scripts/services/mobileClients.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      return DataService.list(mobileclientVersion, {namespace: namespace})
+        .then(function(clients) {
+          return clients.by("metadata.name");
+        });
+    };
+
+    var filterExcluded = function(serviceInstance, mobileClients) {
+      var serviceName = _.get(serviceInstance, 'metadata.name', '');
+      return _.filter(mobileClients, function(client) {
+        var excludedServices = _.get(client, 'spec.excludedServices') || [];
```
Comment (spadgett) [inline app/scripts/services/mobileClients.js]: `_.includes` can take null / undefined, so `|| []` isn't needed

File: app/scripts/services/mobileClients.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        return _.includes(excludedServices, serviceName);
+      });
+    };
+
+    var filterNotExcluded = function(serviceInstance, mobileClients) {
+      var serviceName = _.get(serviceInstance, 'metadata.name', '');
+      return _.filter(mobileClients, function(client) {
+        var excludedServices = _.get(client, 'spec.excludedServices') || [];
```
Comment (spadgett) [inline app/scripts/services/mobileClients.js]: `|| []` not needed

File: app/scripts/services/mobileClients.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        return !_.includes(excludedServices, serviceName);
+      });
+    };
+
+    var removeFromExcluded = function(mobileClient, serviceInstance, context) {
+      var serviceName = _.get(serviceInstance, 'metadata.name', '');
+      _.remove(mobileClient.spec.excludedServices, function(service) {
```
Comment (spadgett) [inline app/scripts/services/mobileClients.js]: Call `angular.copy` before modifying the mobile client.

File: app/styles/_mobile-clients.less [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  .mobile-client {
+    height: 60px;
+    margin-bottom: 1em;
+    overflow: hidden;
+    white-space: nowrap;
+    text-overflow: ellipsis;
+
+    .client-icon {
+     vertical-align: bottom;
```
Comment (spadgett) [inline app/styles/_mobile-clients.less]: indentation

File: app/styles/_mobile-clients.less [hunk 9]
```diff
@@ -0,0 +1,41 @@
+.mobile-clients {
```
Comment (spadgett) [inline app/styles/_mobile-clients.less]: Please keep the declarations sorted alphabetically in this file. We use this style guide:

http://codeguide.patternfly.org/#css-declaration-order

File: app/views/directives/add-mobile-client.html [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  <div class="panel-heading">
+    <h3 class="panel-title pull-left">Select Mobile Client to add to this service</h3>
+    <button class="panel-title btn pull-right" type="button" ng-click="ctrl.onClose()">
+      <span aria-hidden="true" class="pficon pficon-close"></span>
+    </button>
+    <div class="clearfix"></div>
+  </div>
+  <div class="panel-body">
+    <div class="list-pf">
+        <div ng-repeat="client in ctrl.clientsWhereExcluded" class="list-pf-item">
```
Comment (spadgett) [inline app/views/directives/add-mobile-client.html]: indentation

File: app/views/directives/add-mobile-client.html [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<div class="panel panel-default">
```
Comment (spadgett) [inline app/views/directives/add-mobile-client.html]: The dialog should follow Patternfly markup and styles.

Here is an example from the console:

https://github.com/openshift/origin-web-console/blob/master/app/views/modals/confirm-save-log.html

File: app/views/directives/add-mobile-client.html [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  <div class="panel-heading">
+    <h3 class="panel-title pull-left">Select Mobile Client to add to this service</h3>
+    <button class="panel-title btn pull-right" type="button" ng-click="ctrl.onClose()">
+      <span aria-hidden="true" class="pficon pficon-close"></span>
```
Comment (spadgett) [inline app/views/directives/add-mobile-client.html]: Missing `sr-only` text.

File: app/views/directives/add-mobile-client.html (hunk 10 above)
Comment (spadgett) [reply app/views/directives/add-mobile-client.html]:
```html
<div ng-repeat="client in ctrl.clientsWhereExcluded try by (client | uid)" class="list-pf-item">
```

File: app/views/directives/add-mobile-client.html [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    </button>
+    <div class="clearfix"></div>
+  </div>
+  <div class="panel-body">
+    <div class="list-pf">
+        <div ng-repeat="client in ctrl.clientsWhereExcluded" class="list-pf-item">
+          <div class="list-pf-container" ng-click="ctrl.addMobileClient(client)">
```
Comment (spadgett) [inline app/views/directives/add-mobile-client.html]: This is not accessible.

File: app/views/directives/add-mobile-client.html [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+              <div class="order-service-details-top">
+                <div class="service-icon">
+                  <span class="fa-4x {{client.metadata.annotations.icon}}"></span>
+                </div>
+                <div class="service-title-area">
+                  <div class="service-title ng-binding">
+                    {{client.spec.name}}
+                  </div>
+                  <div class="order-service-tags ng-scope">
+                    <span class="tag ng-binding ng-scope">
```
Comment (spadgett) [inline app/views/directives/add-mobile-client.html]: Why are we adding `ng-binding` and `ng-scope` classes?

File: app/views/directives/mobile-service-clients.html [hunk 15]
```diff
@@ -0,0 +1,30 @@
+<div class="row mobile-clients">
+  <div class="col-sm-6">
+    <div class="client-title component-label section-label">Mobile Clients</div>
+    <div ng-repeat="mobileClient in $ctrl.filteredClients" class="col-sm-12 col-md-6 mobile-client">
```
Comment (spadgett) [inline app/views/directives/mobile-service-clients.html]: track by (mobileClient | uid)

File: app/views/directives/mobile-service-clients.html [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+      <span class="fa-2x {{mobileClient.metadata.annotations.icon}} client-icon"></span>
```
Comment (spadgett) [inline app/views/directives/mobile-service-clients.html]: aria-hidden="true"

File: app/views/directives/mobile-service-clients.html [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+      <div class="client-details">
+        <a ng-href="{{mobileClient | navigateResourceURL}}">{{mobileClient.spec.name}}</a>
+        <span class="fa fa-trash-o action-button exclude" ng-click="$ctrl.excludeClient(mobileClient)"></span>
```
Comment (spadgett) [inline app/views/directives/mobile-service-clients.html]: This is not accessible.

File: app/views/directives/mobile-service-clients.html [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        <div class="client-type">{{mobileClient.spec.appIdentifier}}</div>
+      </div>
+    </div>
+  </div>
+
+  <div class="col-xs-12 col-sm-6">
+    <button ng-disabled="$ctrl.canAddMobileClient()" class="btn btn-default btn-lg pull-right"
```
Comment (spadgett) [inline app/views/directives/mobile-service-clients.html]: I'm not sure this should have `btn-lg`?

File: app/views/overview/_service-instance-row.html [hunk 19]
```diff
@@ -152,12 +152,22 @@ <h3>
...
+        </div>
+        <div class="row">
+          <div class="col-sm-12">
             <div ng-if="row.serviceClass.spec.externalMetadata.documentationUrl || row.serviceClass.spec.externalMetadata.supportUrl">
               <a ng-if="row.serviceClass.spec.externalMetadata.documentationUrl" ng-href="{{row.serviceClass.spec.externalMetadata.documentationUrl}}" target="_blank" class="learn-more-link">View Documentation <i class="fa fa-external-link" aria-hidden="true"></i></a>
               <a ng-if="row.serviceClass.spec.externalMetadata.supportUrl" ng-href="{{row.serviceClass.spec.externalMetadata.supportUrl}}" target="_blank" class="learn-more-link">Get Support <i class="fa fa-external-link" aria-hidden="true"></i></a>
             </div>
           </div>
         </div>
+        <div ng-if="row.isMobileEnabled && row.hasMobileClients && row.apiObject | isMobileService">
```
Comment (spadgett) [inline app/views/overview/_service-instance-row.html]: You need parens around the filter

```html
<div ng-if="row.isMobileEnabled && row.hasMobileClients && (row.apiObject | isMobileService)">
```

**Comment by spadgett** [comment]
cc @cshinn

**Comment by cshinn** [comment]
@sedroche Looks good! Just a couple suggestions:

For consistency with other products, it would be better to use pficon-delete in place of the current trash can icon.

Is it possible to show a filtered view of the catalog when selecting a client to add to the project?
If it is not, that modal should use the styling of other OpenShift modals. @spadgett can you think of a good example in the code to base this off of?

**Comment by sedroche** [comment]
@cshinn  Cheers for feedback. When we open the modal we are selecting already provisioned clients rather than provisioning a new instance so the modal makes more sense here.

**Comment by cshinn** [comment]
@sedroche right, makes sense. Another option might be to do something like this
group 2

**Comment by sedroche** [comment]
@cshinn Yes we can do that. I would defer to you on this one so am happy to update if you would like?

**Comment by cshinn** [comment]
@sedroche That's excellent! I checked with some other designers, and the dropdown seems like the best way to go.

**Comment by sedroche** [comment]
@cshinn :+1: I will make this change.

**Comment by sedroche** [comment]
@spadgett Thanks for your review. All feedback makes sense and I will add the requested changes.

**Comment by sedroche** [comment]
@cshinn I have updated the design.

#### New delete icon
new-delete-icon

#### New drop down selector
new-dropdown

**Comment by sedroche** [comment]
@spadgett I believe I have addressed your feedback if you can take another look? Thanks

**Comment by sedroche** [comment]
Closing this branch as we have taken a different direction with this UI
